osh/builtin_assign.py

```diff
@@ -25,10 +26,108 @@
   from frontend import arg_def
 
 
+def _PrintVariables(mem, cmd_val, arg, print_flags, readonly = False, exported = False):
+  flag_g = getattr(arg, 'g', None)
+  flag_n = getattr(arg, 'n', None)
+  flag_r = getattr(arg, 'r', None)
+  flag_x = getattr(arg, 'x', None)
+  flag_a = getattr(arg, 'a', None)
+  flag_A = getattr(arg, 'A', None)
+
+  lookup_mode = scope_e.Dynamic
+  if cmd_val.builtin_id == builtin_i.local:
+    if flag_g and not mem.IsGlobalScope():
+      return 1
+    lookup_mode = scope_e.LocalOnly
+  elif flag_g:
+    lookup_mode = scope_e.GlobalOnly
+
+  if len(cmd_val.pairs) == 0:
+    print_all = True
+    cells = mem.GetAllCells(lookup_mode)
+    names = sorted(cells)
+  else:
+    print_all = False
+    names = []
+    cells = {}
+    for pair in cmd_val.pairs:
+      name = pair.lval.name
+      if pair.rval and pair.rval.tag == value_e.Str:
+        name += "=" + cast(value__Str, pair.rval).s
+        names.append(name)
+        cells[name] = None
+      else:
+        names.append(name)
+        cells[name] = mem.GetCell(name, lookup_mode)
+
+  count = 0
+  for name in names:
+    cell = cells[name]
+    if cell is None: continue
+    val = cell.val
+
+    if val.tag == value_e.Undef: continue
+    if readonly and not cell.readonly: continue
+    if exported and not cell.exported: continue
+    if flag_n == '-' and not cell.nameref: continue
+    if flag_n == '+' and cell.nameref: continue
+    if flag_r == '-' and not cell.readonly: continue
+    if flag_r == '+' and cell.readonly: continue
+    if flag_x == '-' and not cell.exported: continue
+    if flag_x == '+' and cell.exported: continue
+    if flag_a and val.tag != value_e.MaybeStrArray: continue
+    if flag_A and val.tag != value_e.AssocArray: continue
+
+    if print_flags:
+      flags = '-'
+      if cell.nameref: flags += 'n'
+      if cell.readonly: flags += 'r'
+      if cell.exported: flags += 'x'
+      if val.tag == value_e.MaybeStrArray:
+        flags += 'a'
+      elif val.tag == value_e.AssocArray:
+        flags += 'A'
+      if flags == '-': flags += '-'
+
+      decl = 'declare ' + flags + ' ' + name
```

Comment:
generally speaking I use the style of creating a `parts = []` array and appending to it, then `print(''.join(parts))`.

But that is not super important now, if it passes the tests it seems fine

Reply:
Thank you! I changed to use that strategy in d3dd9c97.